src/peft/tuners/oft/layer.py

```diff
@@ -0,0 +1,398 @@
+        scaling = self.scaling[adapter_name]
+        rank = self.r[adapter_name]
+        coft = self.coft[adapter_name]
+        eps = self.eps[adapter_name]
+        opt_r = self.oft_r[adapter_name]
+
+        if coft:
+            with torch.no_grad():
+                opt_r.copy_(self._project_batch(opt_r, eps=eps))
+
+        orth_rotate = self._cayley_batch(opt_r * scaling)
+        weight = self._block_diagonal(orth_rotate, rank)
+
+        return weight
+
+    def _cayley_batch(self, data: torch.Tensor) -> torch.Tensor:
+        b, r, c = data.shape
+        # Ensure the input matrix is skew-symmetric
+        skew = 0.5 * (data - data.transpose(1, 2))
+        I = torch.eye(r, device=data.device).unsqueeze(0).expand(b, r, c)
+
+        # Perform the Cayley parametrization
+        Q = torch.bmm(I - skew, torch.inverse(I + skew))
+
+        return Q
+
+    def _block_diagonal(self, oft_r: torch.Tensor, rank: int) -> torch.Tensor:
+        if oft_r.shape[0] == 1:
+            # block share
+            blocks = [oft_r[0, ...] for i in range(rank)]
+        else:
+            blocks = [oft_r[i, ...] for i in range(rank)]
+
+        # Use torch.block_diag to create the block diagonal matrix
+        A = torch.block_diag(*blocks)
+
+        return A
+
+    def _project_batch(self, oft_r, eps=1e-5):
+        # scaling factor for each of the smaller block matrix
+        eps = eps * 1 / torch.sqrt(torch.tensor(oft_r.shape[0]))
+        I = (
+            torch.zeros((oft_r.size(1), oft_r.size(1)), device=oft_r.device, dtype=oft_r.dtype)
+            .unsqueeze(0)
+            .expand_as(oft_r)
+        )
+        diff = oft_r - I
+        norm_diff = torch.norm(oft_r - I, dim=(1, 2), keepdim=True)
+        mask = (norm_diff <= eps).bool()
+        out = torch.where(mask, oft_r, I + eps * (diff / norm_diff))
+        return out
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        previous_dtype = x.dtype
+
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            result = self.base_layer(x, *args, **kwargs)
+        elif self.merged:
+            result = self.base_layer(x, *args, **kwargs)
+        else:
+            result = None
+            weight = None
+            # Execute all the adapters
+            for active_adapter in self.active_adapters:
+                if active_adapter not in self._available_adapters:
+                    continue
+
+                module_dropout = self.module_dropout[active_adapter]
+
+                if (not self.training) or (self.training and torch.rand(1) > module_dropout):
+                    weight = self._get_delta_activations(active_adapter, weight, *args, **kwargs)
+
+            if weight is None:
+                # skip all active_adapter
+                result = self.base_layer(x, *args, **kwargs)
+            else:
+                result = self._forward_from_weight(x, weight, *args, **kwargs)
```

Comment:
Hmm, can we refactor this to always use `self.base_layer(x, *args, **kwargs)` by adding the delta activations on top? This probably requires to move the updates inside the loop, instead of accumulating all the updated before applying them. This may also simplify some code, as I think we can avoid the `prev_weight` part.

The problem is that if we don't call `self.base_layer(...)`, we completely sidestep that layer's `forward` function. This is okay when we know it's just a normal `nn.Linear`, but we want to allow nested adapters, so the base layer could for instance be an `lora.Linear` layer, in which case this would give the wrong result.

Reply:
@BenjaminBossan 
I have refactored `layer.py`.
To achieve this method, I fixed OFT method similar to kohya_ss.

https://github.com/kohya-ss/sd-scripts/blob/95ae56bd22c285ccb2fe5fca96d92f39842bb99b/networks/oft.py#L77-L89